yt_dlp/extractor/urplay.py

```diff
@@ -69,7 +100,7 @@ def _real_extract(self, url):
         urplayer_streams = urplayer_data.get('streamingInfo', {})
 
         for k, v in urplayer_streams.get('raw', {}).items():
-            if not (k in ('sd', 'hd') and isinstance(v, dict)):
+            if not (k in ('sd', 'hd', 'mp3', 'm4a') and isinstance(v, dict)):
                 continue
```

Comment:
Do we even need this check? Are there known keys that fails here?

Reply:
Not sure. I looked at the downloaded pages, and identified the differences, i.e. the other keys used for audio-only formats.

Was this question asked in the first place, when only `'sd', 'hd'` was introduced? What was the motivation back then?

Reply:
This was implemented upstream by @dirkf

Reply:
Rationale: if there is a `v` that isn't a `dict`, it would definitely be a bad idea to crash extracting from it, whereas if no `v` at all gets extracted the extractor will hit `_sort_formats()` and break there.

https://github.com/ytdl-org/youtube-dl/commit/5cb4833f408745135d1b0e178b9a2545a899f2ac

Reply:
I was questioning the `if not (k in ('sd', 'hd', 'mp3', 'm4a')` part

Reply:
So I realised; see above.

Reply:
BTW, there's another section below also listing `'sd', 'hd'` (where `urplayer_data['streamingInfo'].get('sweComplete')` is evaluated). I guess this tries to parse the subtitles' languages, and therefore probably doesn't apply to audio-only formats - but who prevents distribution of subtitles for audio anyway. ;-)

Reply:
> Rationale: if there is a `v` that isn't a `dict`, it would definitely be a bad idea to crash extracting from it, whereas if no `v` at all gets extracted the extractor will hit `_sort_formats()` and break there.
> 
> [ytdl-org/youtube-dl@5cb4833](https://github.com/ytdl-org/youtube-dl/commit/5cb4833f408745135d1b0e178b9a2545a899f2ac)
> 
> Possibly instead of checking for known media format IDs (`sd`, etc), you could have a check for `is_av_format()` in this loop and one for `is_sttl_format()` in the second loop that extracts subtitles.

I don't understand how to implement `is_av_format()` and `is_sttl_format()` - as far as I can tell, the formats are recognized somewhere along the way by the framework - but I can see that both the `'raw'` and the `'sweComplete'` keys contain A/V media formats. And there's some code duplication by doing two iterations (one for A/V, one for subs).

Here's a suggestion on how they could be combined as such:
https://github.com/barsnick/yt-dlp/commits/urplay.08
Relevant code:
https://github.com/barsnick/yt-dlp/blob/790ec83b23b6e8b27e0202e0f1d5f8a4612bb3f3/yt_dlp/extractor/urplay.py#L106

Only caveat: It's still deciding on A/V vs. subs via the format name.

(`'sweComplete'` actually has a language key for the A/V formats, that's why I'm parsing it early, but I couldn't figure out how to attach the language property to the format in `formats.extend(self._extract_wowza_formats())`.)

Reply:
> ('sweComplete' actually has a language key for the A/V formats, that's why I'm parsing it early, but I couldn't figure out how to attach the language property to the format in formats.extend(self._extract_wowza_formats()).)

```py
fmts = self._extract_wowza_formats(...)
for f in fmts:
    f['language'] = ...
formats.extend(fmts)
```

Reply:
While this was never resolved, I've merged the PR as-is for now